example/node_classification.py

```diff
@@ -0,0 +1,237 @@
+        # label for class is missing for temporary CORA dataset in Memgraph lab
+
+        y[i] = nodes[i].properties.get("class")
+
+        if masks[i] == 1:
+            train_mask[i] = 1
+        elif masks[i] == 2:
+            val_mask[i] = 1
+        else:
+            test_mask[i] = 1
+
+    for i in tqdm(range(len(edges))):
+        edge_index[0][i] = edges[i].from_vertex.properties.get("id")
+        edge_index[1][i] = edges[i].to_vertex.properties.get("id")
+        
+    global NO_CLASSES
+    NO_CLASSES = len(set(y))
+    global NO_FEATURES
+    NO_FEATURES = np.shape(x)[1]
+
+    print("Finished converting data.")
+
+    # print("=============================")
+    # print(np.shape(x))
+    # print(np.shape(y))
+    # print(np.shape(edge_index))
+    # print(np.shape(train_mask))
+    # print(np.shape(val_mask))
+    # print(np.shape(test_mask))
+    # print("=============================")
+
+
+    x = torch.tensor(x, dtype=torch.float32)
+    y = torch.tensor(y, dtype=torch.int64)
+    edge_index = torch.tensor(edge_index, dtype=torch.int64) 
+    train_mask = torch.tensor(train_mask, dtype=torch.bool)
+    val_mask = torch.tensor(val_mask, dtype=torch.bool)
+    test_mask = torch.tensor(test_mask, dtype=torch.bool)
+
+    data = Data(x, y, edge_index, train_mask, val_mask, test_mask)
+    return data
+
+
+model = GCN(hidden_channels=HIDDEN_CHANNELS)
+optimizer = torch.optim.Adam(model.parameters(), lr=LEARNING_RATE, weight_decay=WEIGHT_DECAY)
+criterion = torch.nn.CrossEntropyLoss()
+total_no_epochs = 0
+
+@mgp.read_proc
+def set_parameters(
+    hidden_channels: int,
+    lr: float,
+    weight_decay: float
+) -> mgp.Record(hidden_channels=int, lr=float, weight_decay=float):
+    """
+    set parameters of model and optimizer
+
+    optimizer is Adam, criterion is cross entropy loss
+    """
+    model.hidden_channels = hidden_channels
+    optimizer.lr = lr
+    optimizer.weight_decay = weight_decay
+
+    return mgp.Record(hidden_channels=model.hidden_channels, lr=optimizer.lr, weight_decay=optimizer.weight_decay)
+
+@mgp.read_proc
+def load_data(
+    ctx: mgp.ProcCtx,
+    train_ratio: float = 0.8,
+    val_ratio: float = 0.1,
+    test_ratio: float = 0.1
+) -> mgp.Record(no_nodes=int, no_edges=int):
+    """
+    loading data, must be executed before training
+    """
+
+    nodes = list(iter(ctx.graph.vertices))
+    edges = []
+    for vertex in ctx.graph.vertices:
+        for edge in vertex.out_edges:
+            edges.append(edge)
+    
+    global data
+    data = convert_data(nodes, edges, train_ratio, val_ratio, test_ratio)
+    
+
+
+    return mgp.Record(no_nodes=len(nodes), no_edges=len(edges))
+
+
+@mgp.read_proc
+def train(
+    no_epochs: int = 100
+) -> mgp.Record(total_epochs=int, validation=float):
+    """
+    training
+    """
+
+    # i tried to ensure loaded dataset below, but somehow NameError doesn't get caught
+    # try: 
+    #     data
+    # except NameError: print("Load dataset first!")
```

Comment:
I don’t see `data` as a global variable of the module anywhere.

Reply:
line 200.

Reply:
Thanks! Did you make sure to call `load_data` before `train`?

By the way, the clearest way to create global variables is to declare them outside any function and have them in the same place as the constants - you avoid having to think whether the function that creates them has been run.

Reply:
Repaired.